server/storage/backend/verify.go

```diff
@@ -67,3 +69,41 @@ func insideUnittest() bool {
 	stackTraceStr := string(debug.Stack())
 	return strings.Contains(stackTraceStr, "_test.go") && !strings.Contains(stackTraceStr, "tests/")
 }
+
+// VerifyBackendConsistency verifies data in ReadTx and BatchTx are consistent.
+func VerifyBackendConsistency(b Backend, lg *zap.Logger, skipSafeRangeBucket bool, bucket ...Bucket) {
+	verify.Verify(func() {
+		if b == nil {
+			return
+		}
+		if lg != nil {
+			lg.Debug("verifyBackendConsistency", zap.Bool("skipSafeRangeBucket", skipSafeRangeBucket))
+		}
+		b.BatchTx().LockOutsideApply()
+		defer b.BatchTx().Unlock()
+		b.ReadTx().RLock()
+		defer b.ReadTx().RUnlock()
+		for _, bkt := range bucket {
+			if skipSafeRangeBucket && bkt.IsSafeRangeBucket() {
+				continue
+			}
+			unsafeVerifyTxConsistency(b, bkt)
+		}
+	})
+}
+
+func unsafeVerifyTxConsistency(b Backend, bucket Bucket) {
+	dataFromWriteTxn := map[string]string{}
+	b.BatchTx().UnsafeForEach(bucket, func(k, v []byte) error {
+		dataFromWriteTxn[string(k)] = string(v)
+		return nil
+	})
+	dataFromReadTxn := map[string]string{}
+	b.ReadTx().UnsafeForEach(bucket, func(k, v []byte) error {
+		dataFromReadTxn[string(k)] = string(v)
+		return nil
+	})
+	if diff := cmp.Diff(dataFromWriteTxn, dataFromReadTxn); diff != "" {
```

Comment:
Have you confirmed that this would detect all the recently discovered problems?

When I proposed https://github.com/etcd-io/etcd/issues/17126 I was thinking that we should implement a stub for backend that would implement the same methods but without bbolt and keep the data only in memory. With validate enabled, we would send all request to bbolt to both backend and the stub and compare results. 

I think this would be a better long term approach, however your solution should be acceptable assuming it can reproduce previous issues.

Reply:
It can detect https://github.com/etcd-io/etcd/issues/17158. 
For https://github.com/etcd-io/etcd/pull/17263, since it is rare to write multiple keys in 1 tx, there is no existing issues to detect.

Reply:
> I was thinking that we should implement a stub for backend that would implement the same methods but without bbolt and keep the data only in memory

Sound like a good idea. We clear/reset the stub backend before each test, and compare the real backend with the stub at the end of each test. But when the member restarts, all the in-memory data will be lost, unless we also support persistent storage for the stub backend.  It's a separate topic (This PR just verifies the consistency between read TXN and write TXN.), please feel free to raise a separate ticket to track it if you want.

Reply:
> But when the member restarts, all the in-memory data will be lost, unless we also support persistent storage for the stub backend

You are right that it's impossible to validate cross restart without keeping own persistent storage, however we don't need to test durability, just the runtime behavior. If we validate the data before member is restarted, we should trust the data when backend is loaded back after restart.